### ISSUE TITLE
F3 - open definition seems weired after upgrade

### ISSUE THREAD

**Issue by leomayer** [eclipse/wildwebdeveloper#742]
I tried to update my Eclipse like mentioned in #731. I'm not really sure if the upgrade completely succeeded. Nevertheless just want to be on the save side with the annoyance of the new function to `open definitions`.

Running eclipse.buildId=4.21.0.I20210906-0500

When I press F3 - to open the definition I get the attached menu. I understand the menu to some extend - but to some extent not. To my understanding it should be only the second and/or the fourth entry. The rest seems superfluous. If there would be only one entry of that kind - just open this one. 
![Bildschirmfoto von 2021-11-09 09-45-07](https://user-images.githubusercontent.com/2641128/140892727-8bdb15a6-c4ca-4f62-9e79-9d5eb243ce3b.png)

Any ideas? 

BTW: The keybinding seems as mess in the new Eclipse since a binding to CTRL-SHIFT-8 does NOT work with non English keyboards!!!

**Comment by vrubezhny** [comment]
@leomayer Is it an open source project that I can give a try?

Using the same WWD version and some test project I cannot reproduce the first hyperlink that appears in your screenshot.

If I understand correctly the second and fourth entries are coming from the project, while the rest of items are referencing the return type declaration (like `Array`, f.i., like it happens with my test project) of `processDialogData()`. All those are returned from a Language Server that processes the hyperlink requests.

Not sure if it's also your case, but for my test project I also see some duplicated hyperlinks shown.
Cannot reproduce the first one.

**Comment by leomayer** [comment]
No, its not a public project. I enclosed my config details.

The list of entries varies little bit. Most of the time I can only count 4 items - sometimes 3, sometimes as shown in the screenshot. Anyway I would expect that F3 jumps into the fourth item (which is the declaration I'm interested in).

You need some more logs?

[eclispse.config.zip](https://github.com/eclipse/wildwebdeveloper/files/7670217/eclispse.config.zip)

**Comment by vrubezhny** [comment]
> No, its not a public project. I enclosed my config details.

Thanks for the details.
 
> The list of entries varies little bit. Most of the time I can only count 4 items - sometimes 3, sometimes as shown in the screenshot. Anyway I would expect that F3 jumps into the fourth item (which is the declaration I'm interested in).

With F3 it will never jump in case of more than one hyperlink is available - it always show the list of available proposals in such case in order to let a user choose where to jump.

> You need some more logs?
LSP4E requests and adds all the hyperlink proposals for the method definition plus method return type definition - so having 2-4 or more hyperlinks in the result list is something that is quite expected here.
 
The only thing I see unexpected here is that 'Show in System Explorer` proposal.. Probably your installation doesn't have an editor to open a file referenced in a hyperlink - maybe something is recorded to the LS log for this case? (You can enable it in `Windows -> Preferences -> Language Servers -> Logs` Preference page)

 The labels for the most of hyperlinks are looking quite similar to each other, but it's the way the proposal come from a Language Server (LSP4E in theory can modify the labels to make their text to be more short, but then they will continue to be similar in most of cases. Also it may "break" the labels for some LSs). Also there is no  any indication to distinguish the hyperlinks produced for a method (in your example) definition from the ones produced for a type definition - which could probably be useful if added)

**Comment by mickaelistria** [comment]
@vrubezhny Multiple declarations/definitions are unexpected here. Do you confirm the issue comes from the Language Server returning so many results? If so, do you think you could give a try to a more recent version of the typescript-language-server to verify if it's fixed?

**Comment by vrubezhny** [comment]
@mickaelistria Yes, I confirm that multiple definitions come from the LS. Basically, for a method F3 returns at least its declaration plus the declaring type (I think it's return type definition). For the return type the list of hyperlinks may be quite big, as it  can include all the definitions found for a type in project and the LS internals.
![Screenshot from 2021-12-07 19-34-18](https://user-images.githubusercontent.com/620781/145086815-a7e28bd1-a70c-4451-9f65-2191b20b3625.png)

**Comment by leomayer** [comment]
From my point of view I would like to open the definition of a method - same style as in Eclipse for a regular Java class. I have no clue what to do with the return type definition. I haven't used it till before and don't know what to do. 
I simply want to open the definition of a method (or property) - either from the ˋtsˋ for ˋhtmlˋ file. IFFFF I'm interested in all other occurences I would use the Search dialog. 
@vrubezhny  For the ˋShow in System Explorerˋ: Why shouldn't I be able to open the file? At least the fourth entry I can open! What kind of file type are you talking about and how to figure it out?
